src/engine/engine.signer.test.ts

```diff
@@ -64,3 +82,176 @@ describe('addSignerChange', () => {
     expect(subject()).toEqual([signerChange50A, signerChange50B, signerChange, signerChange200]);
   });
 });
+
+describe('mergeSignerMessage', () => {
+  let aliceCustodySigner: EthereumSigner;
+  let aliceDelegateSigner: Ed25519Signer;
+  let aliceRoot: Root;
+  let genericMessageData: { rootBlock: number; username: string; signedAt: number };
+  let aliceSignerAddDelegate: SignerAdd;
+  let aliceSignerRemoveDelegate: SignerRemove;
+
+  // Generate key pair for alice and root message
+  beforeAll(async () => {
+    aliceCustodySigner = await generateEthereumSigner();
+    aliceDelegateSigner = await generateEd25519Signer();
+    aliceRoot = await Factories.Root.create(
+      { data: { rootBlock: 100, username: 'alice' } },
+      { transient: { signer: aliceCustodySigner } }
+    );
+    genericMessageData = {
+      rootBlock: aliceRoot.data.rootBlock,
+      username: 'alice',
+      signedAt: aliceRoot.data.signedAt + 1,
+    };
+    aliceSignerAddDelegate = await Factories.SignerAdd.create(
+      { data: genericMessageData },
+      { transient: { signer: aliceCustodySigner, childSigner: aliceDelegateSigner } }
+    );
+    aliceSignerRemoveDelegate = await Factories.SignerRemove.create(
+      { data: { ...genericMessageData, body: { childKey: aliceDelegateSigner.signerKey } } },
+      { transient: { signer: aliceCustodySigner } }
+    );
+  });
+
+  // Every test should start with a valid signer and root for alice
+  beforeEach(() => {
+    engine._reset();
+  });
+
+  test('fails without a custody address', async () => {
+    // Hack to add root
```

Comment:
after these three are merged, i think we're ready to remove the notion of roots from our system entirely